cpp/examples/parquet_io/parquet_io_multithreaded.cpp

```diff
@@ -0,0 +1,469 @@
+    "                                <write to temp output files and validate: "
+    "yes/no>\n\n");
+  fmt::print(
+    "Available IO source types: FILEPATH, HOST_BUFFER, {}, DEVICE_BUFFER\n\n",
+    fmt::format(fmt::emphasis::bold | fg(fmt::color::green_yellow), "PINNED_BUFFER (Default)"));
+  fmt::print(fg(fmt::color::light_sky_blue),
+             "Note: Provide as many arguments as you like in the above order. Default values\n"
+             "      for the unprovided arguments will be used. All input parquet files will\n"
+             "      be converted to the specified IO source type before reading\n\n");
+}
+
+/**
+ * @brief Function to process comma delimited input paths string to parquet files and/or dirs
+ *        and asynchronously convert them to specified io sources.
+ *
+ * Process the input path string containing directories (of parquet files) and/or individual
+ * parquet files into a list of input parquet files, multiple the list by `input_multiplier`,
+ * make sure to have at least `thread_count` files to satisfy at least file per parallel thread,
+ * and asynchronously convert the final list of files to a list of `io_source` and return.
+ *
+ * @param paths Comma delimited input paths string
+ * @param input_multiplier Multiplier for the input files list
+ * @param thread_count Number of threads being used in the example
+ * @param io_source_type Specified IO source type to convert input files to
+ * @param stream CUDA stream to use
+ *
+ */
+std::vector<io_source> extract_input_sources_async(std::string const& paths,
+                                                   int32_t input_multiplier,
+                                                   int32_t thread_count,
+                                                   io_source_type io_source_type,
+                                                   rmm::cuda_stream_view stream)
+{
+  // Get the delimited paths to directory and/or files.
+  std::vector<std::string> const delimited_paths = [&]() {
+    std::vector<std::string> paths_list;
+    std::stringstream strstream{paths};
+    std::string path;
+    // Extract the delimited paths.
+    while (std::getline(strstream, path, char{','})) {
+      paths_list.push_back(path);
+    }
+    return paths_list;
+  }();
+
+  // List of parquet files
+  std::vector<std::string> parquet_files;
+  std::for_each(delimited_paths.cbegin(), delimited_paths.cend(), [&](auto const& path_string) {
+    std::filesystem::path path{path_string};
+    // If this is a parquet file, add it.
+    if (std::filesystem::is_regular_file(path)) {
+      parquet_files.push_back(path_string);
+    }
+    // If this is a directory, add all files in the directory.
+    else if (std::filesystem::is_directory(path)) {
+      for (auto const& file : std::filesystem::directory_iterator(path)) {
+        if (std::filesystem::is_regular_file(file.path())) {
+          parquet_files.push_back(file.path().string());
+        }
```

Comment:
Should we report any sub-directories found telling them they will not be included?

Reply:
Not a bad idea. Though we also don't make any effort to ensure the files we are including here are indeed parquet files.